### DISCUSSION TITLE
Easy way to register custom cookies

### DISCUSSION THREAD

**Discussion by robinsowell** [ExpressionEngine/ExpressionEngine#2872] [Feature Requests]
There are two ways for ee to see a cookie as 'registered'.  It's native to core or an add-on (native or third party) addon.setup.php file- see https://docs.expressionengine.com/latest/development/addon-setup-php-file.html#cookies.
 
EE goes through the add-ons on install, and if they say they have cookies, it adds them to the exp_cookie_settings database table.  What's in there are the registered cookies, and their consent type is specified.
 
Any time something uses the native EE cookie setting methods, EE checks those cookie settings to see if the cookie is allowed based on a person's consent settings.  BUT if you are setting a cookie that is NOT in the exp_cookie_settings, it's not registered.  So it will log a dev warning 

> A cookie (some_cookie_name) is being sent without being properly registered, and does not meet cookie compliance policies. Register this cookie appropriately in your addon.setup.php file.
 
And there is currently no way to get custom user defined cookies in there except via the addon.setup.php file.  

What you really want a way to enter custom cookies via the control panel.  Without that, they just aren't going to be 'registered' and you don't have a way to say 'this is a functional cookie' or such like.  That, or maybe allow it in the config.php file- I can see doing overrides that way.  

BUT anyway, it's a feature request, because there's no way to get them in there now.  Well, you could manually add them to the table- I think that would work.  But it's not awesome.  We really need a feature.

**Comment by intoeetive** [comment]
I am not sure about the exact situation here, but ExpressionEngine itself does not provide a template tag to set cookies. So you'll HAVE to use an add-on to set cookie. If the add-on allows setting arbitrary cookie, people could still use `ee('CookieRegistry')->registerFunctionality($cookie_name)` and `ee('CookieRegistry')->registerCookieSettings(CookieSetting $cookie)` 

Docs PR added: https://github.com/ExpressionEngine/ExpressionEngine-User-Guide/pull/734
